### ISSUE TITLE
Cannot send EIP-1559 transaction even when forcing type: 2

### ISSUE THREAD

**Issue by agro1986** [trufflesuite/truffle#4275]
- [ x ] I've asked for help in the [Truffle Gitter](http://gitter.im/Consensys/truffle) before filing this issue, but nobody helped

---------------------------

## Issue

I cannot send EIP-1559 transaction even when forcing type: 2

## Steps to Reproduce

These are the versions in `package.json`

```
"web3": "^1.5.2"
"@truffle/hdwallet-provider": "^1.5.0"
```

Initialize wallet:

```
const provider = new HDWalletProvider(mnemonic, alchemyUrl);
const web3 = new Web3(provider);
```

Then send tx.

1. Doesn't specify anything related to gas:

```
web3.eth.sendTransaction({
        from: account,
        to: toAccount,
        value: sendAmount,
    })
```

2. Specify tx type and various fee per gas settings:

```
web3.eth.sendTransaction({
        from: account,
        to: toAccount,
        value: sendAmount,
        type: 2,
        maxFeePerGas: gasPrice,
        maxPriorityFeePerGas: tip,
    })
```

Both result the same (legacy tx)

## Expected Behavior

EIP-1559 tx created

## Actual Results

Legacy tx created: https://rinkeby.etherscan.io/tx/0x9eeb2345d0d1453896e15eaa8743b74ba44a3a7c9864c66ba49569e990488dc7

## Environment

* Operating System: docker node:14.17.4-alpine3.11
* Ethereum client: alchemy endpoint
* Truffle version (`truffle version`): "@truffle/hdwallet-provider": "^1.5.0" (I don't use the full truffle suite)
* node version (`node --version`): v14.17.4
* npm version (`npm --version`): 6.14.14

**Comment by XWJACK** [comment]
The hdwallet package needs to be updated. 

https://github.com/ChainSafe/web3.js/issues/4245#issuecomment-900881301

**Comment by agro1986** [comment]
@XWJACK thanks. Do you mean that they (the hdwallet-provider guys) need to update it? Or do you mean that I have to update hdwallet-provider to the latest version? Because I'm on the latest released hdwallet-provider version.

**Comment by XWJACK** [comment]
The latest version also does not implement EIP-1559.

This package needs to be merged #4273 first. Then replace `Transaction` with `FeeMarketEIP1559Transaction`.

**Comment by agro1986** [comment]
@davidmurdoch that would be awesome! Where can I know when such special versions are released?

**Comment by davidmurdoch** [comment]
We'll comment on this issue to follow up.

**Comment by eggplantzzz** [comment]
I just published @truffle/hdwallet-provider@eip1559-beta. This dist tag release includes the work from #4273 as well as the work in https://github.com/trufflesuite/truffle/pull/4273.

If you get a chance to try this and have issues then let us know!

**Comment by walruzperil** [comment]
> I just published @truffle/hdwallet-provider@eip1559-beta. This dist tag release includes the work from #4273 as well as the work in #4273.
> 
> If you get a chance to try this and have feedback then let us know!

Тry use 
const web3 = new Web3(providerEngine);
web3.eth.sendTransaction({
    from: "0xb9b7080422a2a1AeaDFb29FDEC545090ed58b31f",
    to: "0x2F50cd2fB35A3f667f6BC0Ea77EF6ff32aF2B9Db", 
    value: web3.utils.toWei("0.01", "ether"), 
    type: 0x02,
    maxFeePerGas: 0x02,
    maxPriorityFeePerGas: 0x02,

}
C:\Users\k1tt1\node_modules\rlp\dist\index.js:61
        throw new Error('invalid remainder');
        ^

Error: invalid remainder

**Comment by agro1986** [comment]
@eggplantzzz Also got error on the beta branch:
```
/Users/agro1986/test/web3-eip1599-test/node_modules/rlp/dist/index.js:61
        throw new Error('invalid remainder');
        ^

Error: invalid remainder
    at Object.decode (/Users/agro1986/test/web3-eip1599-test/node_modules/rlp/dist/index.js:61:15)
    at Object.exports.defineProperties (/Users/agro1986/test/web3-eip1599-test/node_modules/ethereumjs-tx/node_modules/ethereumjs-util/dist/index.js:719:18)
    at new Transaction (/Users/agro1986/test/web3-eip1599-test/node_modules/ethereumjs-tx/es5/index.js:112:13)
    at /Users/agro1986/test/web3-eip1599-test/node_modules/@trufflesuite/web3-provider-engine/subproviders/nonce-tracker.js:62:18
    at /Users/agro1986/test/web3-eip1599-test/node_modules/@trufflesuite/web3-provider-engine/index.js:220:9
    at /Users/agro1986/test/web3-eip1599-test/node_modules/async/internal/withoutIndex.js:9:16
    at replenish (/Users/agro1986/test/web3-eip1599-test/node_modules/async/internal/eachOfLimit.js:66:17)
    at /Users/agro1986/test/web3-eip1599-test/node_modules/async/internal/eachOfLimit.js:71:9
    at eachLimit (/Users/agro1986/test/web3-eip1599-test/node_modules/async/eachLimit.js:43:36)
    at /Users/agro1986/test/web3-eip1599-test/node_modules/async/internal/doLimit.js:9:16
    at end (/Users/agro1986/test/web3-eip1599-test/node_modules/@trufflesuite/web3-provider-engine/index.js:217:5)
    at Object.callback (/Users/agro1986/test/web3-eip1599-test/node_modules/@trufflesuite/web3-provider-engine/subproviders/provider.js:19:5)
    at /Users/agro1986/test/web3-eip1599-test/node_modules/web3-providers-ws/lib/index.js:114:45
    at Array.forEach (<anonymous>)
    at WebsocketProvider._onMessage (/Users/agro1986/test/web3-eip1599-test/node_modules/web3-providers-ws/lib/index.js:102:69)
    at W3CWebSocket._dispatchEvent [as dispatchEvent] (/Users/agro1986/test/web3-eip1599-test/node_modules/yaeti/lib/EventTarget.js:115:12)
```

**Comment by XWJACK** [comment]
```js
var tx = TransactionFactory.fromSerializedData(Buffer.from(ethUtil.stripHexPrefix(rawTx), 'hex'))
```

https://github.com/trufflesuite/provider-engine/blob/d0a1f0a917a0c94d15778f6d5ff09470d71565f3/subproviders/nonce-tracker.js#L62

**Comment by walruzperil** [comment]
> ```js
> var tx = TransactionFactory.fromSerializedData(Buffer.from(ethUtil.stripHexPrefix(rawTx), 'hex'))
> ```
> 
> https://github.com/trufflesuite/provider-engine/blob/d0a1f0a917a0c94d15778f6d5ff09470d71565f3/subproviders/nonce-tracker.js#L62

Received an error 
```
var tx = TransactionFactory.fromSerializedData(Buffer.from(ethUtil.stripHexPrefix(rawTx), 'hex'))
                 ^

ReferenceError: TransactionFactory is not defined
    at @trufflesuite\web3-provider-engine\subproviders\nonce-tracker.js:62:18
```

**Comment by eggplantzzz** [comment]
@XWJACK Are you saying that we will need to update provider-engine as well? At some point we will need to swap out that library for a more current one.

**Comment by XWJACK** [comment]
> @XWJACK Are you saying that we will need to update provider-engine as well? At some point we will need to swap out that library for a more current one.

Fix at https://github.com/trufflesuite/provider-engine/pull/10 and verify it in kovan.

https://kovan.etherscan.io/tx/0xc9526d2b196f437ea6e0b1daee6664b9c8fa053ee179d2db7542084fa0902467

**Comment by walruzperil** [comment]
Trying in Rinkiby:
UnhandledPromiseRejectionWarning: Error: Transaction was not mined within 750 seconds, please make sure your transaction was properly sent. Be aware that it might still be mined!
Simple code:
`web3.eth.sendTransaction({
from: "0x7B318D38624B39e750D7acb48925818198139478",
to: "0x00f6c020353276f4BcFaC63D70608466af45eA8D",

value: web3.utils.toWei("0.01", "ether"),
type: 0x02,
maxFeePerGas: 0x02,
maxPriorityFeePerGas: 0x02,

})`

Maybe need more parameters?

**Comment by eggplantzzz** [comment]
I re-published @truffle/hdwallet-provider@eip1559-beta with the work you did on @trufflesuite/provider-engine, thanks! I tested it with Ropsten and was getting another error: "nonce too low". I'm not sure I understand why it is erroring but I'm not sure I have time to look into it now. Are you able to successfully migrate using @truffle/hdwallet-provider@eip1559-beta @XWJACK?

**Comment by XWJACK** [comment]
> @XWJACK, I re-published @truffle/hdwallet-provider@eip1559-beta with the work you did on @trufflesuite/provider-engine, thanks! However, I tested it with Ropsten by trying to migrate a toy project and got another error: "nonce too low". I'm not sure I understand why it is erroring but I don't think I have time to look into it now. Are you able to successfully migrate using @truffle/hdwallet-provider@eip1559-beta @XWJACK?

It may be that the newly transaction has not been executed by the current script. So the current nonce is not updated. I have been running on the mainnet network without errors. But there are a few things to pay attention to. We need to format as a hexadecimal value or get another error with `throw new Error(Cannot convert string to buffer. toBuffer only supports 0x-prefixed hex strings and this string was given: ${v});`. I will fix it in the next update.
```
{
  maxPriorityFeePerGas: '0x3b9aca00',
  maxFeePerGas: '0xab5d04c00',
}
```

**Comment by XWJACK** [comment]
> Trying in Rinkiby:
> UnhandledPromiseRejectionWarning: Error: Transaction was not mined within 750 seconds, please make sure your transaction was properly sent. Be aware that it might still be mined!
> Simple code:
> `web3.eth.sendTransaction({
> from: "0x7B318D38624B39e750D7acb48925818198139478",
> to: "0x00f6c020353276f4BcFaC63D70608466af45eA8D",
> 
> value: web3.utils.toWei("0.01", "ether"),
> type: 0x02,
> maxFeePerGas: 0x02,
> maxPriorityFeePerGas: 0x02,
> 
> })`
> 
> Maybe need more parameters?

If you specify `maxFeePerGas` and `maxPriorityFeePerGas`, the transaction will automatically set `type: 0x2`. 

> ` Transaction was not mined within 750 seconds` this error is that you have gave a low `maxFeePerGas`

**Comment by jahvi** [comment]
Can confirm this works for me as expected as well, tested on rinkeby:

```json
{
  "from": "0x73257cE508dBB94Ff78C27381FddD42FB7266Ba4",
  "value": "50000000000000000",
  "gas": "0x2dd4d",
  "maxFeePerGas": "0x1a13b862a",
  "maxPriorityFeePerGas": "0x12a05f200",
  "nonce": 105
}
```

https://rinkeby.etherscan.io/tx/0xed3107ad936bc50e089d1cb8b5318442da7344afdf3c24ff304f392299987e0b

**Comment by agro1986** [comment]
@eggplantzzz confirmed working on rinkeby with the latest commit (1.5.1-alpha.1)

```
web3.eth.sendTransaction({
    from: account,
    to: zeroAddress,
    value: new web3.utils.BN(1),
    gasLimit: 21000,
    maxFeePerGas: new web3.utils.BN(web3.utils.toWei('250', 'gwei')),
    maxPriorityFeePerGas: new web3.utils.BN(web3.utils.toWei('4', 'gwei')),
  })
```

https://rinkeby.etherscan.io/tx/0x80b45cfdc773c1506f93aa2b0a365dd8faa608b633140f05cbba100ab54900ef

**Comment by jahvi** [comment]
It worked perfectly for me on my limited use case.

**Comment by agro1986** [comment]
@eggplantzzz yes, I've only tested it for sending ETH though. I hope it can be merged soon so I can deploy contracts on mainnet using EIP1559.

**Comment by eggplantzzz** [comment]
I think this should have been fixed and we forgot to close this. Can anyone confirm/deny that this is fixed on the latest version of Truffle (v5.6.8)?

**Comment by haltman-at** [comment]
As best I can tell, this is no longer occurring.  I'm going to close this.
